### PR TITLE
Optimization

### PR DESCRIPTION
**Pull request by CharlieS1103**
Fixed the PR!

### CODE REVIEW

File: extension.js [hunk 1]
```diff
@@ -294,3 +294,135 @@ const getParamsFromGithubRaw = (url) => {
     const installedExtensions = getLocalStorageDataFromKey(LOCALSTORAGE_KEYS.installedExtensions, []);
     installedExtensions.forEach((extensionKey) => initializeExtension(extensionKey));
 })();
+const ITEMS_PER_REQUEST = 100;
+async function storeThemes() {
+    const BLACKLIST = await Blacklist();
```
Comment (theRealPadster) [inline extension.js:299]: In `storeExtensions` and `storeThemes`, it calls the `Blacklist()` twice. We could likely just call it once somewhere else and have it cached.

File: extension.js [hunk 2]
```diff
@@ -294,3 +294,135 @@ const getParamsFromGithubRaw = (url) => {
...
+            if (extensions) {
+                addToSessionStorage(extensions);
+                await sleep(10000);
+            }
+
+        }
+    }
+}
+// This function is used to fetch manifest of a theme and return it
+async function fetchThemes(contents_url, branch) {
```
Comment (theRealPadster) [inline extension.js:361]: Can we rename this to `fetchThemeManifest` or something? The name is confusing with `storeThemes`.

File: extension.js [hunk 3]
```diff
@@ -294,3 +294,135 @@ const getParamsFromGithubRaw = (url) => {
...
+        }
+        return null;
+    }
+    catch (err) {
+        // console.warn(contents_url, err);
+        return null;
+    }
+}
+// This function is used to fetch manifest of an extension and return it
+async function fetchExtensions(contents_url, branch) {
```
Comment (theRealPadster) [inline extension.js:383]: Can we rename this to `fetchExtensionManifest` or something? The name is confusing with `storeExtensions`.

File: extension.js [hunk 4]
```diff
@@ -294,3 +294,135 @@ const getParamsFromGithubRaw = (url) => {
...
+    const filteredResults = {
+        ...allRepos,
+        items: allRepos.items.filter(item => !BLACKLIST.includes(item.html_url)),
+    };
+    return filteredResults;
+}
+async function storeExtensions() {
+    const BLACKLIST = await Blacklist();
+    let url = `https://api.github.com/search/repositories?q=${encodeURIComponent("topic:spicetify-extensions")}&per_page=${ITEMS_PER_REQUEST}`;
+    const allRepos = await fetch(url).then(res => res.json()).catch(() => []);
```
Comment (theRealPadster) [inline extension.js:314]: This only fetches the first 100 repos returned. It should loop somehow until there are no results left. Something like how `loadPage()` works in `index.js` might work. Same goes for `storeThemes()`. 

```js
const currentPage = requestPage > -1 && requestPage ? requestPage : 1;
// Sets the amount of items that have thus been fetched
const soFarResults = ITEMS_PER_REQUEST * (currentPage - 1) + pageOfRepos.page_count;
const remainingResults = pageOfRepos.total_count - soFarResults;

// If still have more results, return next page number to fetch
console.log(`Parsed ${soFarResults}/${pageOfRepos.total_count} extensions`);
if (remainingResults > 0) return currentPage + 1; // There are more results. currentPage + 1 is the next page to fetch. 
else console.log("No more extension results");
```

File: extension.js [hunk 5]
```diff
@@ -294,3 +294,135 @@ const getParamsFromGithubRaw = (url) => {
...
+async function Blacklist() {
+    const url = "https://raw.githubusercontent.com/CharlieS1103/spicetify-marketplace/main/blacklist.json";
+    const jsonReturned = await fetch(url).then(res => res.json()).catch(() => { });
+    return jsonReturned.repos;
+}
+(async function initializePreload() {
+
+    // Begin by getting the themes and extensions from github
+    const extensionsArray = await storeExtensions();
+    const themesArray = await storeThemes();
```
Comment (theRealPadster) [inline extension.js:333]: Maybe run them concurrently. 
```suggestion
    const [extensionsArray, themesArray] = await Promise.all([
        storeExtensions(), 
        storeThemes(),
    ]);
```

Comment (theRealPadster) [inline extension.js:333]: These could also be named something like "extensionReposArray", to be more clear.

File: extension.js [hunk 6]
```diff
@@ -294,3 +294,135 @@ const getParamsFromGithubRaw = (url) => {
...
+    } else if (type == "extension") {
+        for (const repo of array.items) {
+            let extensions = await fetchExtensions(repo.contents_url, repo.default_branch);
+            if (extensions) {
+                addToSessionStorage(extensions);
+                await sleep(10000);
+            }
+
+        }
+    }
```
Comment (theRealPadster) [inline extension.js:358]: Minor cleanup. 
```suggestion
    for (const repo of array.items) {
        const data = (type === "theme")
            ? await fetchThemes(repo.contents_url, repo.default_branch)
            : await fetchExtensions(repo.contents_url, repo.default_branch);
        if (data) {
            addToSessionStorage(data);
            await sleep(10000);
        }
    }
```

File: extension.js [hunk 7]
```diff
@@ -294,3 +294,135 @@ const getParamsFromGithubRaw = (url) => {
...
+        parsed.push(item);
+        window.sessionStorage.setItem(key, JSON.stringify(parsed));
+
+    });
+
+}
+// This function is used to sleep for a certain amount of time
+function sleep(ms) {
+    return new Promise(resolve => setTimeout(resolve, ms));
+}
```
Comment (theRealPadster) [inline extension.js:428]:
```suggestion
}

```

File: index.js [hunk 8]
```diff
@@ -615,8 +615,15 @@ async function getRepos(page = 1) {
  * @returns The manifest object
  */
 async function getRepoManifest(user, repo, branch) {
-    const url = `https://raw.githubusercontent.com/${user}/${repo}/${branch}/manifest.json`;
+    if (window.sessionStorage.getItem(user + "-" + repo)) {
+        return JSON.parse(window.sessionStorage.getItem(user + "-" + repo));
```
Comment (theRealPadster) [inline index.js:619]: This makes it only grab from sessionStorage once. 
```suggestion
    const sessionStorageItem = window.sessionStorage.getItem(`${user}-${repo}`);
    if (sessionStorageItem) {
        return JSON.parse(sessionStorageItem);
```

File: index.js [hunk 9]
```diff
@@ -778,3 +791,7 @@ async function getThemeRepos(page = 1) {
 
     return filteredResults;
 }
+function getUnloadableManifests() {
+    const unloadableManifests =  window.sessionStorage.getItem("noManifests");
+    return  JSON.parse(unloadableManifests);
+}
```
Comment (theRealPadster) [inline index.js:802]:
```suggestion
}

```

File: index.js [hunk 10]
```diff
@@ -778,3 +791,7 @@ async function getThemeRepos(page = 1) {
 
     return filteredResults;
 }
+function getUnloadableManifests() {
```
Comment (theRealPadster) [inline index.js:799]: This function (and the session storage query from `getRepoManifests`) could be replaced by a single `getSessionStorageValue` function like the one from `index.js`. 

```js
/**
 * Get localStorage data (or fallback value), given a key
 * @param {string} key The localStorage key
 * @param {any} fallback Fallback value if the key is not found
 * @returns The data stored in localStorage, or the fallback value if not found
 */
const getLocalStorageDataFromKey = (key, fallback) => {
    const str = localStorage.getItem(key);
    if (!str) return fallback;

    const obj = JSON.parse(str);
    return obj;
};
```

File: index.js [hunk 11]
```diff
@@ -615,8 +615,15 @@ async function getRepos(page = 1) {
...
+    if (window.sessionStorage.getItem(user + "-" + repo)) {
+        return JSON.parse(window.sessionStorage.getItem(user + "-" + repo));
+    }
 
+    const url = `https://raw.githubusercontent.com/${user}/${repo}/${branch}/manifest.json`;
+    const unloadable = getUnloadableManifests();
+    if ( unloadable.includes(url)) {
+        return null;
+    }
     return await fetch(url).then(res => res.json()).catch(() => null);
```
Comment (theRealPadster) [inline index.js:627]: This should probably cache the value to sessionStorage as well, after it's go it, to reduce calls.

File: index.js [hunk 12]
```diff
@@ -698,11 +705,17 @@ async function fetchRepoExtensions(contents_url, branch, stars) {
...
 async function fetchThemes(contents_url, branch, stars) {
     try {
+        let manifests;
         const regex_result = contents_url.match(/https:\/\/api\.github\.com\/repos\/(?<user>.+)\/(?<repo>.+)\/contents/);
         // TODO: err handling?
         if (!regex_result || !regex_result.groups) return null;
         let { user, repo } = regex_result.groups;
-        let manifests = await getRepoManifest(user, repo, branch);
+        if (window.sessionStorage.getItem(user + "-" + repo)) {
+            manifests = JSON.parse(window.sessionStorage.getItem(user + "-" + repo));
```
Comment (theRealPadster) [inline index.js:719]: Same optimization comment as `getRepoManifest`

File: index.js [hunk 13]
```diff
@@ -631,11 +638,16 @@ async function getRepoManifest(user, repo, branch) {
...
     try {
         // TODO: use the original search full_name ("theRealPadster/spicetify-hide-podcasts") or something to get the url better?
+        let manifests;
         const regex_result = contents_url.match(/https:\/\/api\.github\.com\/repos\/(?<user>.+)\/(?<repo>.+)\/contents/);
         // TODO: err handling?
         if (!regex_result || !regex_result.groups) return null;
         const { user, repo } = regex_result.groups;
-        let manifests = await getRepoManifest(user, repo, branch);
+        if (window.sessionStorage.getItem(user + "-" + repo)) {
+            manifests = JSON.parse(window.sessionStorage.getItem(user + "-" + repo));
```
Comment (theRealPadster) [inline index.js:647]: Same optimization comment as `getRepoManifest`

File: index.js [hunk 14]
```diff
@@ -698,11 +710,17 @@ async function fetchRepoExtensions(contents_url, branch, stars) {
...
+        let manifests;
         const regex_result = contents_url.match(/https:\/\/api\.github\.com\/repos\/(?<user>.+)\/(?<repo>.+)\/contents/);
         // TODO: err handling?
         if (!regex_result || !regex_result.groups) return null;
         let { user, repo } = regex_result.groups;
-        let manifests = await getRepoManifest(user, repo, branch);
+        if (window.sessionStorage.getItem(user + "-" + repo)) {
+            manifests = JSON.parse(window.sessionStorage.getItem(user + "-" + repo));
+        } else {
+            manifests = await getRepoManifest(user, repo, branch);
```
Comment (theRealPadster) [inline index.js:721]: This is the same thing the `extension.js` is caching, I think? So we could cache the results here as well?

File: index.js [hunk 15]
```diff
@@ -631,11 +638,16 @@ async function getRepoManifest(user, repo, branch) {
...
+        let manifests;
         const regex_result = contents_url.match(/https:\/\/api\.github\.com\/repos\/(?<user>.+)\/(?<repo>.+)\/contents/);
         // TODO: err handling?
         if (!regex_result || !regex_result.groups) return null;
         const { user, repo } = regex_result.groups;
-        let manifests = await getRepoManifest(user, repo, branch);
+        if (window.sessionStorage.getItem(user + "-" + repo)) {
+            manifests = JSON.parse(window.sessionStorage.getItem(user + "-" + repo));
+        } else {
+            manifests = await getRepoManifest(user, repo, branch);
```
Comment (theRealPadster) [inline index.js:649]: This is the same thing the `extension.js` is caching, I think? So we could cache the results here as well?

**Review by theRealPadster — CHANGES_REQUESTED** [review]
Added my notes. Looking pretty good, but added a few tweaks, and there are a couple changes that should probably be made. Thanks for tackling this issue! :)

File: extension.js (hunk 1 above)
Comment (CharlieS1103) [reply extension.js:299]: @theRealPadster Done (made a call in initalizePreload to add blacklist to session storage)

File: index.js [hunk 16]
```diff
@@ -580,7 +580,7 @@ class Grid extends react.Component {
  * @param {number} page The query page number
  * @returns Array of search results (filtered through the blacklist)
  */
-async function getRepos(page = 1) {
+async function getExtensionRepos(page = 1) {
```
Comment (theRealPadster) [inline index.js:583]: Let's call this guy `fetchExtensionRepos` to keep it the same as `fetchThemeRepos`

File: index.js [hunk 17]
```diff
@@ -615,16 +615,20 @@ async function getRepos(page = 1) {
  * @returns The manifest object
  */
 async function getRepoManifest(user, repo, branch) {
-    if (window.sessionStorage.getItem(user + "-" + repo)) {
-        return JSON.parse(window.sessionStorage.getItem(user + "-" + repo));
+    const sessionStorageItem = window.sessionStorage.getItem(`${user}-${repo}`);
+    const failedSessionStorageItems = window.sessionStorage.getItem(`noManifests`);
```
Comment (theRealPadster) [inline index.js:619]: It's just a string. 
```suggestion
    const failedSessionStorageItems = window.sessionStorage.getItem("noManifests");
```

File: index.js [hunk 18]
```diff
@@ -796,7 +795,15 @@ async function getThemeRepos(page = 1) {
...
+function addToSessionStorage(items, key) {
+    if (!items || items == null) return;
+    items.forEach(item => {
+        if (!key) key = `${items.user}-${items.repo}`;
+        // If the key already exists, it will append to it instead of overwriting it
+        const existing = window.sessionStorage.getItem(key);
+        const parsed = existing ? JSON.parse(existing) : [];
+        parsed.push(item);
+        window.sessionStorage.setItem(key, JSON.stringify(parsed));
+
```
Comment (theRealPadster) [inline index.js:807]: Random empty line. 
```suggestion
```

**Review by theRealPadster — APPROVED** [review]
It's probably fine now

**Pull request merged**